arroyo/dlq.py

```diff
@@ -17,11 +34,23 @@ class InvalidMessage(Exception):
     should not be retried. It will be placed a DLQ if one is configured.
 
     It can be raised from the submit, poll or join methods of any processing strategy.
+
+    Once a filtered message is forwarded to the next step, `needs_commit` should be set to False,
+    in order to prevent multiple filtered messages from being forwarded for a single invalid message.
     """
 
     def __init__(self, partition: Partition, offset: int) -> None:
         self.partition = partition
         self.offset = offset
+        self.needs_commit = True
+
+
+class InvalidMessageOutOfOrder(Exception):
+    """
+    Fatal exception. Strategies are not permitted to raise invalid messages out of order.
```

Comment:
How would this work on the multi process strategy during the processing phase (that can happen out of order) ? 
Is the parallel processor catching exceptions and then raising them in order?

Reply:
I imagine this could work in a similar way to https://github.com/getsentry/arroyo/blob/4786cb0b5f160a0f9809667cdaf683c8b057ad2c/arroyo/processing/strategies/run_task_with_multiprocessing.py#L205. Once we have processed the entire batch we can sort the invalid messages by offset and start raising them one at at time in order.